### PR TITLE
Patient characteristic codes

### PR DESCRIPTION
**Pull request by losborne**
Fix so PatientCharacteristic toString will properly display codes.

Pull requests into CQL QDM Patient API require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1625
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Code coverage has not gone down and all code touched or added is covered.

branch | coverage
-- | --
bonnie-v2.3 | Statements   : 75.83% ( 1682/2218 )<br>Branches     : 49.26% ( 535/1086 )<br>Functions    : 67.15% ( 327/487 )<br>Lines        : 74.16% ( 1458/1966 )
patient_characteristic_codesbonnie-v2.3 | Statements   : 75.86% ( 1684/2220 )<br>Branches     : 49.36% ( 537/1088 ) <br>Functions    : 67.15% ( 327/487 )<br>Lines        : 74.19% ( 1460/1968 )

- [x] Automated regression test(s) pass N/A does not affect calculations

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA test links: NA
- [x] Justification for using JIRA tests: NA
- [x] JIRA tests have been added to sprint NA

**Reviewer 1:**

Name: @edeyoung 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass NA
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree NA

**Reviewer 2:**

Name: @daco101 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass NA
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree NA

### CODE REVIEW

File: app/assets/javascripts/datatypes/datatype.js.coffee [hunk 1]
```diff
@@ -72,8 +72,12 @@ class CQL_QDM.QDMDatatype
...
     startTime = if @entry?.start_time then "START: #{moment.utc(@entry.start_time, 'X').format('MM/DD/YYYY h:mm A')}\n" else ""
     endTime = if @entry?.end_time then "STOP: #{moment.utc(@entry.end_time, 'X').format('MM/DD/YYYY h:mm A')}\n" else ""
-    # Get code if this datatype has one
-    code = @code() if @_codes
+    # If it is a patient characteristic, use getCode() instead of code()
+    if /PatientCharacteristic/.test(@constructor.name)
+      code = @getCode()
+    else
+      # Get code if this datatype has any
+      code = @code() if @_codes
```
Comment (daco101) [inline app/assets/javascripts/datatypes/datatype.js.coffee]: I think it would be clearer to have something like `else if @_codes: code=@code()` to make it more obvious that code might remain undefined.

File: app/assets/javascripts/datatypes/datatype.js.coffee [hunk 2]
```diff
@@ -72,8 +72,12 @@ class CQL_QDM.QDMDatatype
     # Grab start and end time, format for proper display
     startTime = if @entry?.start_time then "START: #{moment.utc(@entry.start_time, 'X').format('MM/DD/YYYY h:mm A')}\n" else ""
     endTime = if @entry?.end_time then "STOP: #{moment.utc(@entry.end_time, 'X').format('MM/DD/YYYY h:mm A')}\n" else ""
-    # Get code if this datatype has one
-    code = @code() if @_codes
+    # If it is a patient characteristic, use getCode() instead of code()
+    if /PatientCharacteristic/.test(@constructor.name)
+      code = @getCode()
```
Comment (edeyoung) [inline app/assets/javascripts/datatypes/datatype.js.coffee]: why this approach instead of making it so `@code()` works for patient characteristics? Having two separate ways of accessing codes doesn't seem quite right. Looks like perhaps `@getCode` should work for all data criteria, so shifting to that instead of `@code` may also be an option...

Actually (ugh), it looks like `@getCode` as defined in QDMDatatype returns an array and `@getCode` overridden in patient characteristics returns a single code. Perhaps that should be modified. Perhaps `@getCode` in patient characteristics should override and return an array in the patient characteristic, and a `@code` override should be added to return the single code, and then this stuff here doesn't need to change.

Also, kind of weird generally that to test if we should be getting a code through the "public" accessor `@code`, we check that we have it "privately" stored (public/private related to intent rather than reality).

Comment (losborne) [reply app/assets/javascripts/datatypes/datatype.js.coffee]: This approach was taken to solve the problem at hand with minimal change.  

Using `code` instead of `getCode` for `PatientCharacteristic`s as it is implemented now would require adding `@_codes` to them, which doesn't really make sense since they are hard-coded (as you mention below, it doesn't really make sense to check `@_codes` to call `@_code`, so that is a bigger refactor to make all `code()` check for `@_code` themselves).

I agree that `code` and `getCode` could generally use a refactor.  `getCode` should be named `getCodes`, as it is currently a piece of disinformation. I think changing these functions affects the cqm models/conversion, so it will be more time/effort if we want to do that as well.

**Comment by edeyoung** [comment]
This potentially doesn't work with patient characteristic payer. This may be because that is populated more like a typical data criteria rather than the other patient characteristics.

Patient data criteria:

image


Logic displayed results:

image


Excel export:

image
